src/debug/telemetry/telemetry2_thread_info.c

```diff
@@ -0,0 +1,327 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) 2024 Intel Corporation
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <stdio.h>
+#include <soc.h>
+#include <adsp_debug_window.h>
+
+#include "telemetry2_slot.h"
+
+LOG_MODULE_REGISTER(thread_info);
+
+/* to adsp backend header, data to be shared with user space client*/
+#define THREAD_INFO_MAX_THREADS 16
+
+struct thread_info {
+	char name[14];
+	uint8_t stack_usage;	/* precentage */
+	uint8_t cpu_usage;	/* precentage */
+} __packed;
+
+#define THREAD_INFO_VERSION_0_0		0
+
+#define THREAD_INFO_STATE_UNINITIALIZED	0
+#define THREAD_INFO_STATE_BEING_UPDATED	1
+#define THREAD_INFO_STATE_UPTODATE	2
+
+struct thread_info_core {
+	uint8_t state;
+	uint8_t counter;		/* incremented every round */
+	uint8_t load;			/* precentage */
+	uint8_t thread_count;
+	struct thread_info thread[THREAD_INFO_MAX_THREADS];
+} __packed __aligned(64);
+
+struct thread_info_chunk {
+	struct telemetry2_chunk_hdr hdr;
+	uint8_t core_count;
+	uint16_t core_offset[CONFIG_MP_MAX_NUM_CPUS];
+	struct thread_info_core __aligned(64) core[CONFIG_MP_MAX_NUM_CPUS];
+} __packed;
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+static struct previous_counters {
+	uint64_t active;
+	uint64_t all;
+	struct thread_counters {
+		void *tid;
+		uint64_t cycles;
+	} threads[THREAD_INFO_MAX_THREADS];
+} previous[CONFIG_MP_MAX_NUM_CPUS];
+#endif
+
+struct user_data {
+	struct thread_info_core *core_data;
+	int thread_count;
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+	bool stats_valid;
+	uint32_t all_cycles;
+	void *active_threads[THREAD_INFO_MAX_THREADS];
+	struct previous_counters *previous;
+#endif
+};
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+static uint32_t get_cycles(void *tid, k_thread_runtime_stats_t *thread_stats,
+			   struct user_data *ud, const char *name)
+
+{
+	int i;
+
+	ud->active_threads[ud->thread_count] = tid;
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		if (ud->previous->threads[i].tid == tid) {
+			uint32_t cycles = (uint32_t) (thread_stats->execution_cycles -
+						      ud->previous->threads[i].cycles);
+
+			LOG_DBG("%p found at %d (%s %llu)", tid, i,
+				name, thread_stats->execution_cycles);
+			ud->previous->threads[i].cycles = thread_stats->execution_cycles;
+			return cycles;
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		if (ud->previous->threads[i].tid == NULL) {
+			ud->previous->threads[i].tid = tid;
+			ud->previous->threads[i].cycles = thread_stats->execution_cycles;
+			LOG_DBG("%p placed at %d (%s %llu)", tid, i,
+				name, ud->previous->threads[i].cycles);
+			break;
+		}
+	}
+
+	if (i == ARRAY_SIZE(ud->previous->threads))
+		LOG_INF("No place found for %s %p", name, tid);
+
+	return 0;
+}
+
+static uint8_t thread_info_cpu_utilization(struct k_thread *thread,
```

Comment:
It looks like some of this file is RTOS specific and should go in Zephyr as Zephyr APIs.

Reply:
Its just refining the Zephyr thread stats cycle-counter values to produce load information. It requires cached values from previous round, and a generic Zephyr size implementation would become quite complex, requiring caller specific context allocation etc. If there is time for it, I can do it of course, but AFAIK we do not have it ATM.